cmd/sink/expr/cel.go

```diff
@@ -0,0 +1,312 @@
+// if this function returns an error, the Filters struct returned can still be used and will not be nil.
+func NewFilters(path string) (*Filters, error) {
+	var errList []error
+	filters := &Filters{
+		archive: make(map[string]cel.Program),
+		delete:  make(map[string]cel.Program),
+		path:    path,
+	}
+	exists, err := files.PathExists(path)
+	if err != nil {
+		return filters, fmt.Errorf("cannot determine if ConfigMap is mounted at path %s: %s", path, err)
+	}
+	if path == "" || !exists {
+		return filters, fmt.Errorf("cannot create Filters. ConfigMap is not mounted at path %s", path)
+	}
+	filterFiles, err := files.DirectoryFiles(path)
+	if err != nil {
+		return filters, fmt.Errorf(
+			"cannot create Filters. Could not read files created from ConfigMap mounted at path %s: %s",
+			path,
+			err,
+		)
+	}
+	globalArchive, globalDelete, err := getGlobalCelExprs(filterFiles)
+	errList = append(errList, filters.changeGlobalFilters(filterFiles, globalArchive, globalDelete))
+
+	return filters, errors.Join(errList...)
+}
+
+func getGlobalCelExprs(filterFiles map[string]string) (map[string]string, map[string]string, error) {
+	archiveExprs := make(map[string]string)
+	deleteExprs := make(map[string]string)
+	globalFile, exists := filterFiles[globalKey]
+	if !exists {
+		return archiveExprs, deleteExprs, ErrNoGlobal
+	}
+	resourcesBytes, err := os.ReadFile(globalFile)
+	if err != nil {
+		return  archiveExprs, deleteExprs, err
+	}
+	resources := []v1alpha1.KubeArchiveConfigResource{}
+	err = yaml.Unmarshal(resourcesBytes, &resources)
+	if err != nil {
+		return archiveExprs, deleteExprs, err
+	}
+	for _, resource := range resources {
+		gvk := schema.FromAPIVersionAndKind(resource.Selector.APIVersion, resource.Selector.Kind)
+		archiveExprs[gvk.String()] = resource.ArchiveWhen
+		deleteExprs[gvk.String()] = resource.DeleteWhen
+	}
+	return archiveExprs, deleteExprs, nil
+}
+
+// This function must be called when global filters for f have changed. This includes when f is first created.
+func (f *Filters) changeGlobalFilters(filterFiles, globalArchive, globalDelete map[string]string) error {
+	errList := []error{}
+	delete(filterFiles, globalKey)
+	for namespace, filePath := range filterFiles {
+		err := f.createFilters(namespace, filePath, globalArchive, globalDelete)
+		if err != nil {
+			errList = append(errList, err)
+		}
+	}
+	return errors.Join(errList...)
+}
+
+// creates archive and delete filters for namespace from the KubeArchiveConfigResource stored in file. It tries to create
+// all the filters it can even if it encounters errors. Any errors encountered are wrapped together and returned
+func (f *Filters) createFilters(namespace, file string, globalArchive, globalDelete map[string]string) error {
+	resourcesBytes, err := os.ReadFile(file)
+	if err != nil {
+		return err
+	}
+	resources := []v1alpha1.KubeArchiveConfigResource{}
+	err = yaml.Unmarshal(resourcesBytes, &resources)
+	if err != nil {
+		return err
+	}
```

Comment:
This code to read the file and also similar code in `getGlobalCelExprs` can be changed to call a common function to do it:
```
func readResources(filename string) ([]v1alpha1.KubeArchiveConfigResource, error) {
        resources := []v1alpha1.KubeArchiveConfigResource{}
        resourcesBytes, err := os.ReadFile(filename)
        if err != nil {
                return resources, err
        }
        err = yaml.Unmarshal(resourcesBytes, &resources)
        if err != nil {
                return resources, err
        }
        return resources, nil
}       
```
The block of code above could then be changed to simply be:
```
        readResources(file)
```
And similar changes can be made in `getGlobalCelExprs`.

Reply:
done. I created `KACResourceSliceFromFile` and put it in `cmd/sink/k8s/kinds.go`